### PR TITLE
Levenberg-Marquart + SLSQP for Proximity

### LINKED ISSUE #389: Issues with Proximity
Example with non-successful inverse searches:
```python
import splinepy as spp
import numpy as np

# Create a spline
spline = spp.Bezier(
    degrees=[2, 2],
    control_points=np.array([
        [0.0, 0.0],
        [0.2, -1.0],
        [2.0, 0.0],
        [0.0, 0.2],
        [0.2, 0.2],
        [1.5, 0.2],
        [0.0, 1.0],
        [0.2, 2.0],
        [1.0, 1.0]
    ])
)

# Make sure results are always available
solution = spp.utils.data.cartesian_product(
    [np.linspace(0, 1, 10) for _ in range(2)])
queries = spline.evaluate(solution)

# Search inverses
nearest = spline.proximities(
    queries=queries,
    max_iterations=200,
    initial_guess_sample_resolutions=[5, 5]
)

# Output
print(
    f"Proximity failed for points: "
    f"{np.where(np.linalg.norm(solution - nearest, axis=1) > 1e-7)[0]}"
)
```

`Proximity failed for points: [10 21 30 38 41 51 58 60 61 62 63 68 78 81 82 83 85 86 88]`

Increasing the number of samples creates more hits, however, this should always work, imo. Even with only 3 starting points.

Issues:
- [ ] Fix example
- [ ] Clean up unused functions (`GaussWithPivot`)

Check out #386 for more information

### PR DESCRIPTION
**Pull request by jzwar**
# Overview
Adds Levenberg-Marquart and SLSQP for proximity. Current implementation wouldn't work for some situations (see #389). Source for SLSQP is from current scipy's fortran version. To avoid difficulties of fortran-compiler-setup in CI systems, the source is translated to `c` using `f2c`. Then, removed some static variables and add const keywords.

Now, the search looks like this:
```
Initial guess
|
Newton
|
Converged? -- yes -- return
|
No
|
Reset initial guess, max_iterations *= 5
|
SLSQP
|
Converged? -- yes -- return
|
No
|
Reset initial guess, max_iterations *= 5
|
LM
```
Also, SLSQP can incorporate bounds. Nice.

## Addressed Issues
- fixes #389
- Sometimes, newton iterations does not converge
- Sometimes, we needed to sample a lot of points for k-d tree
- Sometimes, we really want to get correct values at any cost.

## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [ ] Added test(s)

### CODE REVIEW

**Comment by j042** [comment]
Do you have scripts/notes of the origin of this implementation?

**Comment by j042** [comment]
never mind, found it in  #389

**Comment by j042** [comment]
> one thing to note in general is that we probably had to call `system.ResetRowOrder();` at each iteration.

On the other hand, it would have iterated through this order and reordered it, so maybe it is irrelevant

**Review by j042 — COMMENTED** [review]
Awesome addition! I think we can set Newton max_iter a bit lower than LM. I've added some discussion points.

File: src/proximity/proximity.cpp [hunk 1]
```diff
@@ -158,13 +176,258 @@ void Proximity::FillLhs(const RealArray_& guess,
...
+        PrepareIterationLevenbergMarquart(data);
+        ComputeStatus(data, false);
+        continue;
+      }
+    }
+
+    // if rho is too big, decrease lambda
+    if (chosen_rho > data.LM.upper_bound) {
+      data.LM.lambda *= 0.5;
+      // i think this should have "continue" here - double check!
```
[the remark this answers is not in the document]
Comment (j042) [reply src/proximity/proximity.cpp:417]: looking at the notes, I guess this no `continue` is intended here

**Comment by j042** [comment]
If you get a chance, I'd appreciate if you can take a look at `proximity.hpp` / `proximity.cpp` @danielwolff1 :)

File: src/proximity/proximity.cpp [hunk 2]
```diff
@@ -158,18 +290,336 @@ void Proximity::FillLhs(const RealArray_& guess,
...
+    // Prepare next step and compute status
+    PrepareIterationNewton(data);
+    ComputeStatus(data, false);
+
+    data.status.stop_iteration = ++i;
+  }
+}
+
+void Proximity::PrepareIterationLevenbergMarquart(SearchData& data) const {
+  // LM requires Jac (pronounced similar to Jacques)
```
Comment (danielwolff1) [inline src/proximity/proximity.cpp:407]: That comment, lol 🤣

File: src/proximity/proximity.cpp [hunk 3]
```diff
@@ -178,107 +628,100 @@ void Proximity::VerboseQuery(
...
-            lhs);
-    convergence_norm = rhs.NormL2();
+    return;
   }
 
-  // before returning, we just need to fill symmetric part of hessian
-  for (int i{}; i < spline_hessian.Shape()[0]; ++i) {
-    for (int j{i + 1}; j < spline_hessian.Shape()[1]; ++j) {
-      std::copy_n(&spline_hessian(i, j, 0), dim, &spline_hessian(j, i, 0));
+  // Newton didn't work. Try LevenbergMarquart
```
Comment (danielwolff1) [inline src/proximity/proximity.cpp:676]:
```suggestion
  // Newton and Slsqp didn't work. Try LevenbergMarquart
```

**Review by danielwolff1 — APPROVED** [review]
I took a look at it and have some minor comments.

I have to admit that I did not check every single line of code in detail, but I took a few hours to check closely and to me it looks good, I didn't spot anything that seems completely wrong. Therefore, I approve! 🚀

File: include/splinepy/proximity/proximity.hpp [hunk 4]
```diff
@@ -66,22 +242,83 @@ class Proximity {
...
   /// @param guess_phys
   /// @param difference
-  void GuessMinusQuery(const RealArray_& guess,
-                       const ConstRealArray_& query,
-                       RealArray_& guess_phys,
-                       RealArray_& difference) const;
+  void GuessMinusQuery(const RealArray& guess,
+                       const ConstRealArray& query,
+                       RealArray& guess_phys,
+                       RealArray& difference) const;
```
[the remark this answers is not in the document]
Comment (j042) [reply include/splinepy/proximity/proximity.hpp]: Done!
I've noticed similar situations with `dJdu`  and `d2Jdu2` so I've changed them too. for `J`, I guess it didn't matter but made it the same as `GuessMinusQuery`. ([see this commit](https://github.com/tataratat/splinepy/pull/394/commits/5923d0d9b2ee15a04d220fe3f69d5ac60bc29d6a))

[the remark this answers is not in the document]
Comment (danielwolff1) [reply include/splinepy/proximity/proximity.hpp]: Great, thanks a lot, I know stuff like this does not anything in terms of the functionality but I personally think that it greatly enhances usability! 😊

**Comment by danielwolff1** [comment]
Thanks for incorporating the changes, from my point of view this can be merged! 🚀

**Review by danielwolff1 — APPROVED** [review]

**Comment by j042** [comment]
Thanks for the review @danielwolff1 !

**Pull request merged**